### ISSUE TITLE
Rendering a Opengl texture

### ISSUE THREAD

**Issue by statefull** [wang-bin/QtAV#498]
Is it possible to pass directly a opengl texture to be rendered? I know that a surface interop can be used to perform the render, but  is there a way to set opengl texture to be rendered?

**Comment by wang-bin** [comment]
use fbo in renderer?

**Comment by statefull** [comment]
yes I'm performing a FBO to render to a texture and I'm wondering if that texture could be set in qtav to be rendered.

**Comment by wang-bin** [comment]
Do you mean how to use QtAV  to renderer the video to a FBO? If yes, you can use `OpenGLVideo` class. qml/QuiclFBORenderer.cpp is an example.

**Comment by statefull** [comment]
No, I have in my code a GL_RGBA texture and I want to know if there is a way to set that texture in QtAV to be rendered. Just use QtAV to render the opengl texture directly.

**Comment by wang-bin** [comment]
attatch your texture to fbo `glFramebufferRenderbuffer` then use `OpenGLVideo` render the video? You may have to use opengl functions directly instead of `QOpenGLFrameBuffer` because there's no function to attach an existing texture. I'm not sure it's right because I'm not familar with fbo.

**Comment by statefull** [comment]
I'm checking for implement a surface interop and I realize that a handler is passed, should I fill that handler (which is a GLuint (a opengl texture))  with the data that I want to be shown?

**Comment by wang-bin** [comment]
surface interop is used by hw decoders to convert the decoded data to opengl textures. those textures are then rendered by opengl to fbo or window. so it's not you need.

**Comment by statefull** [comment]
The problem is that I'm using intel media server for decoding h264 stream and it seems that there is a bug in the vaapi intel driver that crash when I try to use the vaapi of surface interop when try to copy the surface to the GLX surface, so I want to make it programatically by getting the data using vaderiveimage and put it to the QtAV surface (opengl), so is it possible?

**Comment by wang-bin** [comment]
X11 is the default `display` property for vaapi decoder since 1.7. It supports interop with opengl without coping data back from gpu.

**Comment by wang-bin** [comment]
Default `copyMode` is `ZeroCopy` for X11 display. You can change to `GenericCopy` or `OptimizedCopy` to use vagetimage and then rendered by opengl

**Comment by statefull** [comment]
interesting, so is there a way to ensure that zerocopy is being used? or how can I change the GenericCopy.... to use this should I use X11SurfaceInterop I think, right? I would like to measure the performance of each option...

**Comment by wang-bin** [comment]
The default is zerocopy if possible. You can manually change the behavior too https://github.com/wang-bin/QtAV/wiki/Enable-Hardware-Decoding

**Comment by wang-bin** [comment]
You can test in player example, use the decoder setup menu(not setup dialog)

**Comment by statefull** [comment]
could you check the code I'm using...

VADisplay disp;
VASurfaceID surfaceID;
QtAV::vaapi::display_ptr vaapiDisplayPtr;
QtAV::vaapi::surface_ptr *vaapiSurface_ptr;
QtAV::vaapi::X11InteropResource glxInterop;
QtAV::vaapi::InteropResourcePtr *glxInteropPtr;

surface is the surface that I get from the decoder... where the VADISPLAY and VASURFACEID are contained.

 glxInteropPtr = new QtAV::vaapi::InteropResourcePtr(&glxInterop);
_surfaceInterop = QtAV::VideoSurfaceInteropPtr(new QtAV::vaapi::SurfaceInteropVAAPI(*glxInteropPtr));

disp = surface->m_va_display;

vaapiDisplayPtr = QtAV::vaapi::display_ptr(new QtAV::vaapi::display_t(disp));

 vaapiSurface = new QtAV::vaapi::surface_t(frame->GetWidth(),frame->GetHeight(),surfaceID,vaapiDisplayPtr);
 vaapiSurface_ptr = new QtAV::vaapi::surface_ptr(vaapiSurface);

 ((QtAV::vaapi::SurfaceInteropVAAPI *)_surfaceInterop.data())->setSurface(*vaapiSurface_ptr,w,h);

 QRTCVideoFrame f(w, h, QtAV::VideoFormat::Format_NV12);

f.setBytesPerLine(w*4);
f.setMetaData("surface_interop", QVariant::fromValue(_surfaceInterop));
QVariantHash opt;
opt["display"] = "X11";
opt["copyMode"] = "ZeroCopy";
f.setMetaData("display", opt.take("display"));
f.setMetaData("copyMode",opt.take("copyMode"));

this frame is passed to a queue and then it is render... the problem is i'm getting black screen. could NV12 be used?

**Comment by wang-bin** [comment]
`QtAV::vaapi` is for internal use only. Why do you use them in your code? All these are done in QtAV. The `QVariantHash` options is the parameter of `AVPlayer::setOptionsForVideoCodec(options)` and `VideoDecoder::setOptions(options)`. `AVPlayer` is enough for user applications.

Sometimes the format is NV12.

**Comment by wang-bin** [comment]
I see. You have your own decoder and you want to render va surfaces using QtAV, is that right?
For every VideoFrame, you have to create a SurfaceInteropVAAPI and set VideoSurfaceInteropPtr as frame's metadata["surface_interop"]. Only 1 InteropResourcePtr is enough. The format is always `VideoFormat::Format_RGB32` because `vaPutSurface` will convert yuv format to rgb format. Then opengl can render the frame correctly.

**Comment by statefull** [comment]
yes, I think I'm doing that stuff, I create a videoframe and then I add the surface interop in metadata. The problem is that if I use GLX interop the driver returns an error and if I use X11 interop i'm getting black screen. I have checked that the surface has correct data... also I want to know if using interop zeroCopy could be enabled

**Comment by wang-bin** [comment]
If the metadata is set, 0-copy will be automatically enabled.

**Comment by statefull** [comment]
Just a question.. is there a way to get the opengl context QtAV is using?

**Comment by wang-bin** [comment]
`QOpenGLContext::currentContext()`

**Comment by statefull** [comment]
If I set information to the handler (GLuint) used in QtAV which is passed to the surface interop that information is rendered?

**Comment by wang-bin** [comment]
can you show me your code

**Comment by statefull** [comment]
Now, I can see the image that I'm redering here is the code... The problem is that the image refresh is very slow or even sometimes it is not refreshed although the images are begin generated correctly.  The refresh of the image could be related with this function OpenGLVideo::render(const QRectF &target, const QRectF& roi, const QMatrix4x4& transform) ?

fbo->bind();
const QRect drawRect(0, 0, pixmap_width, pixmap_height);
const QSize drawRectSize = drawRect.size();

QOpenGLPaintDevice device(drawRectSize);
QPainter painter;
painter.begin(&device);
painter.setRenderHints(QPainter::Antialiasing | QPainter::HighQualityAntialiasing);

painter.fillRect(drawRect, Qt::blue);

painter.setPen(QPen(Qt::green, 5));
painter.setBrush(Qt::red);
painter.drawEllipse(0, rand()%100, rand()%400, rand()%200);
painter.drawEllipse(rand()%100, 0, rand()%200, rand()%400);

painter.setPen(QPen(Qt::white, 0));
QFont font;
font.setPointSize(24);
painter.setFont(font);
painter.drawText(drawRect, "Hello FBO", QTextOption(Qt::AlignCenter));
painter.end();

fbo->release();

QImage img = fbo->toImage();
img = img.convertToFormat(QImage::Format_RGB32);

//TEX REFERENCES THE QtAV texture Handler
glBindTexture(GL_TEXTURE_2D,tex);

glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, w,h, GL_RGBA, GL_UNSIGNED_BYTE, img.bits());

glBindTexture(GL_TEXTURE_2D,0);

**Comment by wang-bin** [comment]
I don't know what you want to do. If you want to render the va surface, just create a video frame and set the interop object.

**Comment by statefull** [comment]
as i told you I cannt do that because the intel driver has a bug and crash when QtAV want to copy the data from the VA surface to the GLX surface. I getting the data from the VA surface and I want to load that data to the QtAV handler (which is a opengl texture). I'm doing this OK, the problem is that QtAV is not refreshing the image only when I force the window by modifying its size

**Comment by wang-bin** [comment]
Does QtAV player's vaapi decoder+zero copy work?
You can try SurfaceInteropVAAPI::mapToHost if you want to copy

**Comment by wang-bin** [comment]
If you use fbo, you can try your own interop class and implement `createHandle` which returns the ptr to fbo texture. In `map` draw your content. Then no need to convert to qimage and upload texture.

**Comment by statefull** [comment]
I follow your recomendations and now I create a new surface interop and returning the fbo texture as handler... but I have the same problem the image is only shown if a resize the window...

here is the code...

CREATING THE HANDLER

void * SurfaceInteropFBH::createHandle(void *handle, SurfaceType type, const VideoFormat &fmt, int plane, int planeWidth, int planeHeight)
{
    Q_UNUSED(handle);
    Q_UNUSED(type);
    Q_UNUSED(fmt);
    Q_UNUSED(plane);

    fboFormat.setSamples(16);
    fboFormat.setAttachment(QOpenGLFramebufferObject::CombinedDepthStencil);

    if(!fbo) fbo = new QOpenGLFramebufferObject(planeWidth/4,planeHeight,GL_TEXTURE_2D);
    fbo->bind();
    _handler = fbo->texture();
    fbo->release();

   return &_handler;
}

THE MAPPING

void* SurfaceInteropFBH::map(SurfaceType type, const VideoFormat &fmt, void *handle, int plane)
{
    if (!handle)
        return NULL;
    if (!fmt.isRGB())
        return 0;

    if (!m_surface)
        return 0;

    QOpenGLFunctions * func = m_context->functions();

    func->glBindBuffer(GL_ARRAY_BUFFER,0);

    if(!fbo->isBound())
        fbo->bind();



    const QRect drawRect(0, 0, frame_width, frame_height);
    const QSize drawRectSize = drawRect.size();

    QOpenGLPaintDevice device(drawRectSize);
    QPainter painter;
    painter.begin(&device);
    painter.setRenderHints(QPainter::Antialiasing | QPainter::HighQualityAntialiasing);

    painter.fillRect(drawRect, Qt::blue);


    painter.setPen(QPen(Qt::green, 5));
    painter.setBrush(Qt::red);
    painter.drawEllipse(0, rand()%100, rand()%400, rand()%200);
    painter.drawEllipse(rand()%100, 0, rand()%200, rand()%400);

    painter.setPen(QPen(Qt::white, 0));
    QFont font;
    font.setPointSize(24);
    painter.setFont(font);
    painter.drawText(drawRect, "Hello FBO", QTextOption(Qt::AlignCenter));

    painter.end();

    fbo->release();

    return handle;
}

**Comment by wang-bin** [comment]
`map` is called but it's not shown? so strange

**Comment by wang-bin** [comment]
can you try `mpv --hwdec=vaapi file` to see if 0-copy works

**Comment by statefull** [comment]
yes the map is working due to in the texture handle the image is, but the image is shown only if I perform a window resize (like forcing a window refresh). If I'm all the time resizing the window the image is updated.

**Comment by statefull** [comment]
regarding MPV it is working... but QtAV is crashing if we use VAAPI surface interop. 

./mpv --hwdec=vaapi /Downloads/11.h264 
Playing: /Downloads/11.h264
 (+) Video --vid=1 (h264)
libva info: VA-API version 0.35.0
libva info: va_getDriverName() returns 0
libva info: User requested driver 'iHD'
libva info: Trying to open /opt/intel/mediasdk/lib64/iHD_drv_video.so
libva info: Found init function __vaDriverInit_0_32
libva info: va_openDriver() returns 0
This stream has no timestamps!
Making up playback time using 29.970030 FPS.
Seeking will probably fail badly.
Using hardware decoding (vaapi).
VO: [opengl] 352x288 vaapi
V: 00:00:21 (99%)

**Comment by statefull** [comment]
I think there is a problem with this...
For example is always repainting the background
d.out_rect != boundingRect().toRect();

is like there is a problem with sizes or something... any tip that I can check?

**Comment by statefull** [comment]
basically the problem is that when the mapping is done and the texture has the data is like the render is not being notified or something to show the texture...

**Comment by wang-bin** [comment]
I find that vaapi always crash if nvidia gpu is usd. I think there are some problems in my code and maybe it's the reason why you crash.
Are you using QML?

**Comment by wang-bin** [comment]
mapping is only happend when the renderer is going to render the video

**Comment by statefull** [comment]
yes, i'm using qml for show the image

**Comment by wang-bin** [comment]
Which renderer are you using? QQuickItemRenderer or QuickFBORenderer?

**Comment by statefull** [comment]
QuickFBORenderer

**Comment by wang-bin** [comment]
Is `glDrawArrays` in `OpenGLVideo::render()` called when mapping is done?

**Comment by statefull** [comment]
yes, I checked it.

**Comment by wang-bin** [comment]
OK, I will try you code.

**Comment by statefull** [comment]
thanks
